### ISSUE TITLE
SC2154 var is referenced but not assigned not triggered if used within an array

### ISSUE THREAD

**Issue by robstoll** [koalaman/shellcheck#2970]
#### For bugs
- Rule Id (if any, e.g. SC1000):  SC2154
- My shellcheck version (`shellcheck --version` or "online"): 0.10.0
- [ ] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [ ] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

#### For new checks and feature suggestions
- [ ] https://www.shellcheck.net/ (i.e. the latest commit) currently gives no useful warnings about this
- [ ] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/usr/bin/env bash
a=("$non-existing")
echo "${a[@]}"
```

#### Here's what shellcheck currently says:
```
$ shellcheck myscript
No issues detected!
```

#### Here's what I wanted or expected to see:
```
$ shellcheck myscript
 
[Line 4:](javascript:setPosition(4, 7))
a=("$non-existing")
      ^-- [SC2154](https://www.shellcheck.net/wiki/SC2154) (warning): nonExisting is referenced but not assigned.
```

I have not contributed code to shellcheck by now but would be willing to write my first PR if I get some pointers :)

**Comment by koalaman** [comment]
This was a kludge from back before ShellCheck supported associative arrays. At the time, `declare -A x=([y]=z)` would treat `y` as a reference just like , causing a lot of false positives, so the check simply ignored references in arrays.

Parsing of associative arrays has since been fixed, so it was just a matter of removing this kludge. Thanks!
